soc/riscv/riscv-privilege/telink_b91/power.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2022 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ext_driver/ext_pm.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/pm.h>
+#include <zephyr/pm/device.h>
+#include <stimer.h>
+#include <zephyr/zephyr.h>
+
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+#define DT_DRV_COMPAT telink_machine_timer
+
+#define MTIME_REG	DT_INST_REG_ADDR(0)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 8)
+
+#define mticks_to_systicks(mticks)                                                                 \
+	(((uint64_t)(mticks)*SYSTEM_TIMER_TICK_1S) / CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
+#define systicks_to_mticks(sticks)                                                                 \
+	(((uint64_t)(sticks)*CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC) / SYSTEM_TIMER_TICK_1S)
+
+#define SYSTICKS_MAX_SLEEP 0xe0000000
+#define SYSTICKS_MIN_SLEEP 18352
+
+/**
+ * @brief This define converts Machine Timer ticks to B91 System Timer ticks.
+ */
+#define MTIME_TO_STIME_SCALE (SYSTEM_TIMER_TICK_1S / CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
+
+/**
+ * @brief Get Machine Timer Compare value.
+ */
+static uint64_t get_mtime_compare(void)
+{
+	return *(const volatile uint64_t *const)((uint32_t)(MTIMECMP_REG +
+						 (_current_cpu->id * sizeof(uint64_t))));
+}
+
+/**
+ * @brief Get Machine Timer value.
+ */
+static uint64_t get_mtime(void)
+{
+	const volatile uint32_t *const rl = (const volatile uint32_t *const)MTIME_REG;
+	const volatile uint32_t *const rh =
+		(const volatile uint32_t *const)(MTIME_REG + sizeof(uint32_t));
+	uint32_t mtime_l, mtime_h;
+
+	do {
+		mtime_h = *rh;
+		mtime_l = *rl;
+	} while (mtime_h != *rh);
+	return (((uint64_t)mtime_h) << 32) | mtime_l;
+}
+
+/**
+ * @brief Set Machine Timer value.
+ */
+static void set_mtime(uint64_t time)
+{
+	volatile uint32_t *const rl = (volatile uint32_t *const)MTIME_REG;
+	volatile uint32_t *const rh =
+		(volatile uint32_t *const)(MTIME_REG + sizeof(uint32_t));
+
+	*rl = 0;
+	*rh = (uint32_t)(time >> 32);
+	*rl = (uint32_t)time;
+}
+
+/**
+ * @brief PM state set API implementation.
+ */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(substate_id);
+
+	uint32_t tl_sleep_tick = stimer_get_tick();
+	uint64_t current_time = get_mtime();
+	uint64_t wakeup_time = get_mtime_compare();
+
+	if (wakeup_time <= current_time) {
+		LOG_DBG("Sleep Time = 0 or less\n");
+		return;
+	}
+#ifdef CONFIG_PM_DEVICE
+	if (pm_device_is_any_busy()) {
+		return;
+	}
+#endif /* CONFIG_PM_DEVICE */
```

Comment:
this is not where this check should be done. If your system can't go to sleep because a device activity doesn't allow for a particular power state you must use the policy state_lock APIs.

Reply:
in our platform sleep stops all devices. so if any device is in process of its action we can't enter sllep. we cant separate core and devices because sleep mode affect everything.

Reply:
so that means device drivers should lock sleep state/s while they are active.